// victory-native/dist/cartesian/utils/axisDefaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameDefaults = exports.YAxisDefaults = exports.XAxisDefaults = exports.CartesianAxisDefaultProps = void 0;
const react_native_1 = require("react-native");
exports.CartesianAxisDefaultProps = {
    lineColor: "hsla(0, 0%, 0%, 0.25)",
    lineWidth: react_native_1.StyleSheet.hairlineWidth,
    tickCount: 5,
    labelOffset: { x: 2, y: 4 },
    axisSide: { x: "bottom", y: "left" },
    axisScales: { xAxisScale: "linear", yAxisScale: "linear" },
    labelPosition: "outset",
    formatXLabel: (label) => String(label),
    formatYLabel: (label) => String(label),
    labelColor: "#000000",
    ix: [],
    domain: null,
};
exports.XAxisDefaults = {
    lineColor: "hsla(0, 0%, 0%, 0.25)",
    lineWidth: react_native_1.StyleSheet.hairlineWidth,
    tickCount: 5,
    labelOffset: 2,
    axisSide: "bottom",
    yAxisSide: "left",
    labelPosition: "outset",
    formatXLabel: (label) => String(label),
    labelColor: "#000000",
    labelRotate: 0,
};
exports.YAxisDefaults = {
    lineColor: "hsla(0, 0%, 0%, 0.25)",
    lineWidth: react_native_1.StyleSheet.hairlineWidth,
    tickCount: 5,
    labelOffset: 4,
    axisSide: "left",
    labelPosition: "outset",
    formatYLabel: (label) => String(label),
    labelColor: "#000000",
    yKeys: [],
    domain: null,
};
exports.FrameDefaults = {
    lineColor: "hsla(0, 0%, 0%, 0.25)",
    lineWidth: react_native_1.StyleSheet.hairlineWidth,
};
