// victory-native/dist/shared/chartCanvasSizeUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyChartLayoutChange = exports.shouldWarnMissingHeadlessExplicitSize = exports.isChartHeadless = exports.resolveChartCanvasSizeState = exports.getInitialChartCanvasSizeState = void 0;
function getInitialChartCanvasSizeState(explicitSize) {
    return {
        size: explicitSize !== null && explicitSize !== void 0 ? explicitSize : { width: 0, height: 0 },
        hasMeasuredLayoutSize: Boolean(explicitSize),
    };
}
exports.getInitialChartCanvasSizeState = getInitialChartCanvasSizeState;
function resolveChartCanvasSizeState(layoutSizeState, explicitSize) {
    return explicitSize
        ? getInitialChartCanvasSizeState(explicitSize)
        : layoutSizeState;
}
exports.resolveChartCanvasSizeState = resolveChartCanvasSizeState;
function isChartHeadless(headless, explicitSize) {
    return Boolean(headless && explicitSize);
}
exports.isChartHeadless = isChartHeadless;
function shouldWarnMissingHeadlessExplicitSize(headless, explicitSize) {
    return Boolean(headless && !explicitSize);
}
exports.shouldWarnMissingHeadlessExplicitSize = shouldWarnMissingHeadlessExplicitSize;
function applyChartLayoutChange(explicitSize, layout) {
    if (explicitSize) {
        return null;
    }
    return { size: layout, hasMeasuredLayoutSize: true };
}
exports.applyChartLayoutChange = applyChartLayoutChange;
