// victory-native/dist/shared/useChartCanvasSize.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useChartCanvasSize = exports.shouldWarnMissingHeadlessExplicitSize = exports.resolveChartCanvasSizeState = exports.isChartHeadless = exports.getInitialChartCanvasSizeState = exports.applyChartLayoutChange = void 0;
const React = __importStar(require("react"));
const chartCanvasSizeUtils_1 = require("./chartCanvasSizeUtils");
var chartCanvasSizeUtils_2 = require("./chartCanvasSizeUtils");
Object.defineProperty(exports, "applyChartLayoutChange", { enumerable: true, get: function () { return chartCanvasSizeUtils_2.applyChartLayoutChange; } });
Object.defineProperty(exports, "getInitialChartCanvasSizeState", { enumerable: true, get: function () { return chartCanvasSizeUtils_2.getInitialChartCanvasSizeState; } });
Object.defineProperty(exports, "isChartHeadless", { enumerable: true, get: function () { return chartCanvasSizeUtils_2.isChartHeadless; } });
Object.defineProperty(exports, "resolveChartCanvasSizeState", { enumerable: true, get: function () { return chartCanvasSizeUtils_2.resolveChartCanvasSizeState; } });
Object.defineProperty(exports, "shouldWarnMissingHeadlessExplicitSize", { enumerable: true, get: function () { return chartCanvasSizeUtils_2.shouldWarnMissingHeadlessExplicitSize; } });
function useChartCanvasSize({ explicitSize, headless, }) {
    const [size, setSize] = React.useState({ width: 0, height: 0 });
    const [hasMeasuredLayoutSize, setHasMeasuredLayoutSize] = React.useState(false);
    const isHeadless = (0, chartCanvasSizeUtils_1.isChartHeadless)(headless, explicitSize);
    const resolvedChartCanvasSizeState = (0, chartCanvasSizeUtils_1.resolveChartCanvasSizeState)({ size, hasMeasuredLayoutSize }, explicitSize);
    React.useEffect(() => {
        if (!(0, chartCanvasSizeUtils_1.shouldWarnMissingHeadlessExplicitSize)(headless, explicitSize)) {
            return;
        }
        console.warn("`headless` requires `explicitSize`. Falling back to layout-measured rendering.");
    }, [headless, explicitSize]);
    const onLayout = React.useCallback(({ nativeEvent: { layout } }) => {
        const result = (0, chartCanvasSizeUtils_1.applyChartLayoutChange)(explicitSize, layout);
        if (!result) {
            return;
        }
        setHasMeasuredLayoutSize(result.hasMeasuredLayoutSize);
        setSize(result.size);
    }, [explicitSize]);
    return Object.assign(Object.assign({}, resolvedChartCanvasSizeState), { onLayout,
        isHeadless });
}
exports.useChartCanvasSize = useChartCanvasSize;
